### DISCUSSION TITLE
Computed Reactions?

### DISCUSSION THREAD

**Discussion by LandonSchropp** [mobxjs/mobx#2897] [Q&A]
I'm using `mobx-state-router` to control my router state. I want to add a `RoutersParamsStore` that serves to wrap the router's parameters and create usable models for my application. The idea is when you navigate to a route, it automatically constructs the object needed for that route.

Here's what my `RouterParamsStore` looks like:

``` js
import { makeAutoObservable } from "mobx";

import { ExampleStore } from "./example-store";

export class RouterParamsStore {

  router = null

  constructor({ router }) {
    console.log("RouterParamsStore#constructor");
    this.router = router;
    makeAutoObservable(this);
  }

  get example() {
    console.log("RouterParamsStore#example");
    return new ExampleStore({ id: this.router.routerState.params.id });
    return example;
  }
}
```

``` js
import { makeAutoObservable } from "mobx";

export class ExampleStore {

  id = ""
  values = [ "" ]

  constructor(id) {
    makeAutoObservable(this);
    this.id = id;
    console.log("ExampleStore#constructor", this.values[0]);
  }

  get loudValue() {
    return this.value.toUpperCase();
  }

  setValue(value) {
    this.value = value;
  }
}
```

The idea here is to take advantage of MobX's computed caching. Since the router params aren't changing between navigation, the objects will persist and can be used to store the state.

The problem I'm having is that `ExampleStore` needs to access its own computed properties during its construction. I *think* what's happening is the `get example` property is binding to the state inside of `ExampleStore`. When this state changes, the computed property automatically recalcs, and my state gets wiped out.

Is there a way I can tell MobX to only pay attention to a subset of accessed properties in a computed? I'd imagine this would sort of work like a reaction, but with a computed property.

I promise there's a real use case behind this question—this is just an example.

Thanks in advance!

**Comment by LandonSchropp** [comment]
This is my temporary workaround for this problem:

``` js
import { makeAutoObservable, reaction } from "mobx";

import { ExampleStore } from "./example-store";

export class RouterParamsStore {

  router = null
  example

  constructor({ router }) {
    console.log("RouterParamsStore#constructor");

    this.router = router;
    makeAutoObservable(this);

    reaction(
      () => this.router.routerState,
      () => this.loadData()
    );

    this.loadData();
  }

  loadData() {
    let { routeName } = this.router.routerState;

    console.log(routeName);

    if (routeName === undefined) {
      return;
    }

    switch (routeName) {
      case "example":
        this.example = this.loadExample();
        break;
      default:
        throw new Error(`Unrecognized route ${ routeName }`);
    }
  }

  loadExample() {
    console.log("RouterParamsStore#example");
    return new ExampleStore({ id: this.router.routerState.params.id });
  }
}
```

I'm really not in love with this solution. It makes move verbose, less readable, and it's preemptively evaluating the models rather than lazily loading them.

**Comment by urugator** [comment]
Computeds returning complex objects are troublesome, I would avoid using the cache as persistent state holder. You often want to keep the object synchronized rather than recreated or you often want it to survive unobservation, do some async work etc.
You can try to experiment with [`untracked`](https://mobx.js.org/api.html#untracked), [`observe`](https://mobx.js.org/intercept-and-observe.html#observe), [`onBecomeObserved`](https://mobx.js.org/api.html#onbecomeobserved), `_allowStateChanges` (allows mutations, like syncing, in computed).
However if I may suggest, if you need to synchronize stores, create something lazily etc, do it as part of a single action. In your case I would have a `navigate` action that simply handles anything that should happen during navigation - updating router store, creating or reloading page specific store, disposing other stores etc. Any "reactive" solution becomes tangled mess that's hard to reason about - any mutation can cause a cascade of other arbitrary logic basically anywhere inside your app.

**Reply by LandonSchropp** [reply]
Makes sense. Thanks for the suggestion!
